### PR TITLE
Script to automatically add licence headers

### PR DESCRIPTION
**Pull request by mariantalla**
This PR adds a `ensure-boilerplate.sh` script; when run, it will add copyright/licence headers to applicable files.

Where a valid licence header is already present, the script will not update or otherwise change it. Similarly for files that it is configured to `skip`.

As discussed with [sig-contributor-experience](https://docs.google.com/document/d/1qf-02B7EOrItQgwXFxgqZ5qjW0mtfu5qkYIF1Hl4ZLI/edit#) this is copied from [pivotal-k8s/kubernetes-contrib-tools](https://github.com/pivotal-k8s/kubernetes-contrib-tools) - we'll deprecate the latter once this PR is merged.

/cc @hoegaarden

### CODE REVIEW

**Comment by nikhita** [comment]
To follow the same conventions as scripts in hack/, can we rename this to `update-boilerplate.sh` instead of `ensure-boilerplate.sh`?

Also, I am a little confused about how repo-infra is being used for k/k right now. k/k has it's own boilerplate verification: https://github.com/kubernetes/kubernetes/blob/master/hack/verify-boilerplate.sh and https://github.com/kubernetes/kubernetes/tree/master/hack/boilerplate.

/cc @cblecker @spiffxp

File: verify/boilerplate/boilerplate.py [hunk 1]
```diff
@@ -47,10 +46,24 @@
...
 args = parser.parse_args()
 
 verbose_out = sys.stderr if args.verbose else open("/dev/null", "w")
 
+default_skipped_dirs = ['Godeps', 'third_party', '_gopath', '_output', '.git', 'cluster/env.sh',
+                        "vendor", "test/e2e/generated/bindata.go", "hack/boilerplate/test",
+                        "pkg/generated/bindata.go"]
+
+# list all the files that contain 'DO NOT EDIT', but are not generated
+default_skipped_not_generated = ['hack/build-ui.sh', 'hack/lib/swagger.sh',
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/boilerplate/boilerplate.py]: As of now, we are not sure where `repo-infra` is actually used. We'd like to see it being used everywhere, also in `k/k`.

As we started of using the script from `k/k` we kept the defaults as they were in the `k/k` repo. However, we introduced an option to override those via a config file in other repos.
Maybe we can add more (clearer) docs about exactly that.

**Comment by hoegaarden** [comment]
- [As said](https://github.com/kubernetes/repo-infra/pull/82#discussion_r216986134), we are not sure where `repo-infra` is used and where it is not. The boilerplate stuff is different between in `k/k` and in `repo-infra` (before our change). That might be, because it has not been updated in a while or because `k/k` never had used `repo-infra`. But ideally, at least from our point of view, `k/k` should use `repo-infra` and should have some mechanism to automatically update to new versions of `repo-infra`.
- We went for `ensure-boilerplate,sh` instead of `update-boilerplate.sh` because it does not really update anything if the boilerplate is already there, it only adds the boilerplate on files which do not have it. Having said that, if you think `update-boilerplate.sh` is a better name we are happy to rename the thing.

File: verify/boilerplate/boilerplate.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 
-# Copyright 2015 The Kubernetes Authors.
+# Copyright 2018 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/boilerplate/boilerplate.py]: Sorry ... this came from testing the boilerplate code on the boilerplate code itself :) Will revert that.

File: verify/verify-boilerplate.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
-# Copyright 2014 The Kubernetes Authors.
+# Copyright 2018 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/verify-boilerplate.sh]: [1]

**Comment by nikhita** [comment]
> We went for ensure-boilerplate,sh instead of update-boilerplate.sh because it does not really update anything if the boilerplate is already there, it only adds the boilerplate on files which do not have it. Having said that, if you think update-boilerplate.sh is a better name we are happy to rename the thing.

No strong opinions. `ensure-boilerplate` sgtm.

I'm not aware of the history between repo-infra and k/k. I'll defer to @ixdy @spiffxp @cblecker  on that one.

**Comment by hoegaarden** [comment]
Hello @nikhita @ixdy @spiffxp @cblecker .

Do you have any input on that? We'd like to move that forward and see if we can even use that in `k/k`. Any thoughts?

**Comment by cblecker** [comment]
I like the idea, but python isn't my area of expertise.

WDTY @ixdy @spiffxp? Maybe we could rope @fejta into helping review if you're good with the idea :)

File: verify/boilerplate/boilerplate.py [hunk 4]
```diff
@@ -128,25 +171,34 @@ def file_passes(filename, refs, regexs):
...
-                'cluster/env.sh', 'vendor', 'test/e2e/generated/bindata.go',
-                'repo-infra/verify/boilerplate/test', '.glide']
 
-def normalize_files(files):
+def read_config_file(conf_path):
+    try:
+        with open(conf_path) as json_data_file:
+            return json.load(json_data_file)
+    except ValueError:
+        raise
```
Comment (tao12345666333) [inline verify/boilerplate/boilerplate.py]: we should raise error and the reasons.

**Comment by nikhita** [comment]
cc @justaugustus @dims @swinslow 

This relates to our licensing efforts in Kubernetes.

**Comment by hoegaarden** [comment]
Is there anything we can do to get this merged?

Or even more importantly: is there a way to track which repos actually use this repo and therefore assess if its even relevant?

**Comment by nikhita** [comment]
> is there a way to track which repos actually use this repo and therefore assess if it's even relevant?


I believe the verify script for license headers is not used anywhere currently. k/k has its own verify script but the plan is to start using this repo in the future.

**Comment by ixdy** [comment]
> As said, we are not sure where repo-infra is used and where it is not. The boilerplate stuff is different between in k/k and in repo-infra (before our change). That might be, because it has not been updated in a while or because k/k never had used repo-infra. But ideally, at least from our point of view, k/k should use repo-infra and should have some mechanism to automatically update to new versions of repo-infra.

Right, I don't have a good answer for how we can easily vendor scripts from repo-infra and keep them up to date; just copying the scripts from repo-infra into the various other repos is not great. Anyone have any good suggestions?

**Comment by hoegaarden** [comment]
What if this repo had a script which would update itself via `git subtree merge` that could be called from within the parent repo (`k/k`).

I'd be up for experimenting with that in `kubernetes-sigs/testing_frameworks`.

**Comment by hoegaarden** [comment]
Can people (@ixdy, @mikedanese as main contributors & and everyone else really) let us know where (in `k/k`? somewhere else?) and how (as subtree, as submodule, ...) you use this repo? That might inform how we can make this repo easily consumable and update-able from within other repos.

Meanwhile I am experimenting with an update / verify script for the situation when `repo-infra` was `subtree`ed into other repos [over here](https://gist.github.com/hoegaarden/67aa50eeea7bf015dcfc7a51ab083d15).

**Comment by ixdy** [comment]
Several repos vendor kazel using the standard Go vendoring workflows.

A few repos also vendor repo-infra as a Bazel WORKSPACE dependency.

There might be some repos that copy the scripts from `verify/`, but I'm not sure how to track those.

File: verify/boilerplate/boilerplate.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ixdy) [reply verify/boilerplate/boilerplate.py]: +1 to `.boilerplate.json` instead of `boilerplate.json`.

[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/boilerplate/boilerplate.py]: - We will remove the `k/k` defaults and come up with a set of defaults that are reasonable for a broad range of repos
- We will switch to using `.boilerplate.json` for the config file

File: verify/README.md [hunk 5]
```diff
@@ -33,9 +33,45 @@ script:
...
+  "not_generated_files_to_skip" : [
+    "some/file",
+    "some/other/file.something"
+  ]
+}
+```
+Currently supported settings are
+- `dirs_to_skip`  
+  A list of directories which is excluded when checking or adding the headers
+- `not_generated_files_to_skip`
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/README.md]: I'll leave it for now until someone comes up with a better name. IIRC we took the name from a variable or so in the boilerplate scripts when we "forked" of from `k/k` and thought about naming back then too, but couldn't come up with anything much better ...

File: boilerplate.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "dirs_to_skip" : [
+    "vendor/",
+    "boilerplate/test_with_config_file/",
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply boilerplate.json]: Yes and no.

No: [In the code](https://github.com/kubernetes/repo-infra/pull/82/files#diff-af23c75c1db73639c3c17326acec0df9R192) a file is ignored by substring match, e.g. when `dirs_to_skip` includes "to_ingore" and there is a file with the path `/some/random/path/to/a/file/to_ignore/a/sample.txt` this file will ignored, even if the `dirs_to_skip` is not the absolute path, nor the relative path from repo root. So prepending the `verify/` it is not *mandatory*.

Yes: It is probably best to be more specific and set it to `verify/boilerplate/test_with_config_file/`.

File: verify/boilerplate/boilerplate_test.py [hunk 7]
```diff
@@ -14,39 +14,195 @@
...
+
+        output = sorted(sys.stdout.getvalue().split())
+
+        sys.stdout = old_stdout
+
+        self.assertItemsEqual(
+            output, ['././fail.go', '././fail.py', '././fail.sh'])
+
+    def test_read_config(self):
+        config_file = "./test_with_config_file/boilerplate.json"
```
[the remark this answers is not in the document]
Comment (hoegaarden) [reply verify/boilerplate/boilerplate_test.py]: Will check on that.

**Comment by spiffxp** [comment]
Needs rebase

**Comment by spiffxp** [comment]
... automation should really be removing the needs-rebase label automatically on push

File: verify/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (spiffxp) [reply verify/README.md]: I feel silly for asking, but why not `files_to_skip`?

**Comment by fejta** [comment]
/lifecycle stale

Any chance you're going to come back to this? It looks cool and super useful!

**Comment by hoegaarden** [comment]
@fejta I think all the issues and comments are addressed. So I think it is done from our perspective.

**Comment by fejta** [comment]
Cool! Meaning you're waiting on a review?

/remove-lifecycle stale

**Comment by hoegaarden** [comment]
Yes, that'd be great.

**Comment by nikhita** [comment]
Bumping this. Looks like this needs more reviewers (I'm way out of touch with python)

**Comment by nikhita** [comment]
Can we merge and iterate here?

We require donated repos to have this boilerplate header for all their files. Adding the header to each file in a repo is extremely cumbersome, this script would be immensely helpful in such cases.

**Pull request closed without merging**